File: src/ClaudeStatusBar/Data/ChildProcessSpec.cs
namespace ClaudeStatusBar.Data;

/// <summary>
/// The child command line ClaudeCliChannel launches, made injectable so tests can
/// point the channel at a fake child (tests/FakeClaudeChild) instead of the real
/// claude.exe. ResolveExePath is a Func, not a string, because the review requires
/// the path be re-resolved on EVERY launch attempt (Codex review High #1): Claude
/// Code auto-update replaces the binary at that path, so a path cached once at
/// process start can go stale across a long-running supervised session.
///
/// EnvironmentOverrides (docs/multi-account.md) is how one account's child is
/// pointed at a config directory other than Claude Code's own default: setting
/// CLAUDE_CONFIG_DIR is exactly what "N accounts = N config directories = N child
/// processes" means in practice. It is null for the default account so its child's
/// environment is simply inherited unmodified -- CLAUDE_CONFIG_DIR is left for the
/// child to resolve itself exactly as claude.exe always has.
/// </summary>
public sealed record ChildProcessSpec(
    Func<string> ResolveExePath,
    string Arguments,
    string WorkingDirectory,
    IReadOnlyDictionary<string, string>? EnvironmentOverrides = null)
{
    public static ChildProcessSpec Default() => ForAccount(slot: "default", configDir: null);

    /// <summary>
    /// Builds the spec for one account slot (docs/multi-account.md). `slot` is an opaque
    /// per-account identifier the caller controls (StatusBarApplicationContext uses "default"
    /// for accounts[0] and the account's index for every other one); it only affects the
    /// working directory, so two accounts never share one claude.exe session directory.
    ///
    /// `configDir` null means "Claude Code's own default login": CLAUDE_CONFIG_DIR is left
    /// unset in EnvironmentOverrides entirely (not set to the resolved default path) so the
    /// child inherits this process's own environment exactly as it always has -- if the user
    /// already has CLAUDE_CONFIG_DIR set for their whole session, that keeps working unchanged.
    /// A non-null configDir is a second/third account and gets CLAUDE_CONFIG_DIR pinned
    /// explicitly so it can never drift onto whatever the default account happens to resolve to.
    /// </summary>
    public static ChildProcessSpec ForAccount(string slot, string? configDir)
    {
        string agentRoot = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "ClaudeStatusBar", "agent");
        string workingDirectory = string.Equals(slot, "default", StringComparison.Ordinal)
            ? agentRoot // unchanged from the pre-multi-account layout: the sole/first account keeps its existing folder
            : Path.Combine(agentRoot, slot);

        Dictionary<string, string>? overrides = null;
        if (!string.IsNullOrEmpty(configDir))
            overrides = new Dictionary<string, string> { ["CLAUDE_CONFIG_DIR"] = configDir };

        return new ChildProcessSpec(
            ResolveExePath: () => ResolveClaudeExe(
                Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
                Environment.GetEnvironmentVariable("PATH"),
                File.Exists),
            Arguments: "-p --input-format stream-json --output-format stream-json --verbose",
            WorkingDirectory: workingDirectory,
            EnvironmentOverrides: overrides);
    }

    /// <summary>
    /// The native installer's location first (%USERPROFILE%\.local\bin\claude.exe), then the
    /// first claude.exe on PATH. Only a real .exe qualifies: an npm install's claude.cmd shim
    /// runs node under cmd.exe and can't be supervised the same way. When nothing is found the
    /// native path is returned, so the launch failure names where Claude Code was expected.
    /// </summary>
    public static string ResolveClaudeExe(string userProfile, string? pathVariable, Func<string, bool> exists)
    {
        string nativeInstall = Path.Combine(userProfile, ".local", "bin", "claude.exe");
        if (exists(nativeInstall)) return nativeInstall;

        foreach (string entry in (pathVariable ?? "").Split(Path.PathSeparator,
                     StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries))
        {
            string candidate;
            try { candidate = Path.Combine(entry.Trim('"'), "claude.exe"); }
            catch (ArgumentException) { continue; }
            if (exists(candidate)) return candidate;
        }
        return nativeInstall;
    }
}

File: src/Mac/Sources/ClaudeQuotaCore/ChildProcessSpec.swift
import Foundation

/// The child command line `ControlChannel` launches, made injectable so tests can point the
/// channel at a fake child instead of the real `claude`. Mirrors
/// `src/Windows/Data/ChildProcessSpec.cs`; the differences are all measured facts from
/// docs/mac-port.md, "Phase 1 — evidence measured on macOS".
public struct ChildProcessSpec: Sendable {
    /// Re-resolved on EVERY launch attempt, never cached (Codex review High #1 on the Windows
    /// side). On macOS the reason is sharper than on Windows: `~/.local/bin/claude` is a
    /// *symlink* into `~/.local/share/claude/versions/<version>`, and auto-update repoints it.
    /// A path resolved once at app start goes stale the first time Claude Code updates itself.
    public var resolveExecutablePath: @Sendable () -> String
    public var arguments: [String]
    public var workingDirectory: String
    /// Environment entries forced for this child.
    public var environmentOverrides: [String: String]
    /// Environment entries removed from the inherited environment before launching. See
    /// `forAccount` for why this is not merely cosmetic.
    public var environmentRemovals: Set<String>
    /// Where this child records its pid + start time, so a later run can clean it up if it is
    /// left behind (containment mechanism 3 in `ChildProcess`).
    public var pidFileURL: URL?

    public init(
        resolveExecutablePath: @escaping @Sendable () -> String,
        arguments: [String],
        workingDirectory: String,
        environmentOverrides: [String: String] = [:],
        environmentRemovals: Set<String> = [],
        pidFileURL: URL? = nil
    ) {
        self.resolveExecutablePath = resolveExecutablePath
        self.arguments = arguments
        self.workingDirectory = workingDirectory
        self.environmentOverrides = environmentOverrides
        self.environmentRemovals = environmentRemovals
        self.pidFileURL = pidFileURL
    }

    public static let protocolArguments = [
        "-p", "--input-format", "stream-json", "--output-format", "stream-json", "--verbose",
    ]

    /// Claude Code reads this *in preference to* `CLAUDE_CONFIG_DIR` when deciding which keychain
    /// namespace a login belongs to — see docs/mac-port.md §2 for the decompiled selector. If the
    /// user happens to have it exported, every account child would resolve the same namespace and
    /// the accounts would collide, despite each having its own `CLAUDE_CONFIG_DIR`.
    public static let secureStorageVariable = "CLAUDE_SECURESTORAGE_CONFIG_DIR"
    public static let configDirVariable = "CLAUDE_CONFIG_DIR"

    public static func `default`() -> ChildProcessSpec {
        forAccount(slot: "default", configDirectory: nil)
    }

    /// Builds the spec for one account slot (docs/multi-account.md). `slot` is an opaque
    /// per-account identifier the caller controls; it decides the working directory and the pid
    /// record, so two accounts never share one `claude` session directory.
    ///
    /// Spawning this child creates a real Claude Code session and fires SessionStart hooks, so
    /// the working directory is a dedicated agent folder, never a user repo.
    public static func forAccount(slot: String, configDirectory: String?) -> ChildProcessSpec {
        let agentRoot = agentRootURL()
        let workingDirectory = slot == "default"
            ? agentRoot.path
            : agentRoot.appendingPathComponent(slot).path

        var overrides: [String: String] = [:]
        var removals: Set<String> = []
        if let configDirectory, !configDirectory.isEmpty {
            // Phase 1 §2: Claude Code derives the login's Keychain service name from
            // `sha256(NFC(configDir))`, so the *spelling* of this path is part of the account's
            // identity. A login performed under one spelling is invisible to a child started
            // under another, which presents as an account that is silently logged out.
            let canonical = canonicalConfigDirectory(configDirectory)
            overrides[configDirVariable] = canonical
            // Pin the secure-storage selector to the same directory rather than leaving whatever
            // the user's environment holds. Setting it to the same canonical path is a no-op for
            // the resulting namespace (both spell `-sha256(canonical)[0..8]`), so this only ever
            // removes an inherited surprise.
            overrides[secureStorageVariable] = canonical
        } else {
            // The DEFAULT account deliberately inherits the user's environment unmodified: if
            // they have set CLAUDE_CONFIG_DIR for their whole session, that keeps working, and
            // whatever secure-storage directory they chose goes with it.
            removals = []
        }

        return ChildProcessSpec(
            resolveExecutablePath: { resolveClaudeExecutable() },
            arguments: protocolArguments,
            workingDirectory: workingDirectory,
            environmentOverrides: overrides,
            environmentRemovals: removals,
            pidFileURL: agentRoot.appendingPathComponent("\(slot).pid"))
    }

    public static func applicationSupportDirectory() -> URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? URL(fileURLWithPath: NSHomeDirectory()).appendingPathComponent("Library/Application Support")
        return base.appendingPathComponent("ClaudeStatusBar")
    }

    public static func agentRootURL() -> URL {
        applicationSupportDirectory().appendingPathComponent("agent")
    }

    /// The one spelling of a config directory this app ever uses: absolute, symlinks resolved,
    /// no trailing slash, Unicode-normalised the same way Claude Code normalises it (NFC).
    /// See docs/mac-port.md §2, "the config-dir path must be canonical and stable".
    ///
    /// `resolvingSymlinksInPath` only resolves components that exist, so a not-yet-created
    /// directory still normalises deterministically -- the caller creates it before logging in.
    public static func canonicalConfigDirectory(_ path: String) -> String {
        let expanded = (path as NSString).expandingTildeInPath
        let url = URL(fileURLWithPath: expanded).standardizedFileURL.resolvingSymlinksInPath()
        var result = url.path
        while result.count > 1 && result.hasSuffix("/") { result.removeLast() }
        return result.precomposedStringWithCanonicalMapping   // NFC
    }

    /// The native installer's location first, then the two package managers, then `PATH`.
    ///
    /// Phase 1 §1: on this machine only `~/.local/bin/claude` exists, as a symlink to a real
    /// Mach-O binary. Unlike Windows there is no shim to screen out. `PATH` is only a backstop:
    /// a GUI app launched from Finder or as a login item does not inherit the shell's `PATH`, so
    /// in production the explicit paths are the real mechanism. When nothing is found the native
    /// path is returned, so the launch failure names where Claude Code was expected.
    public static func resolveClaudeExecutable(
        home: String = NSHomeDirectory(),
        pathVariable: String? = ProcessInfo.processInfo.environment["PATH"],
        isExecutable: (String) -> Bool = { FileManager.default.isExecutableFile(atPath: $0) }
    ) -> String {
        let native = (home as NSString).appendingPathComponent(".local/bin/claude")
        let candidates = [native, "/opt/homebrew/bin/claude", "/usr/local/bin/claude"]
        for candidate in candidates where isExecutable(candidate) { return candidate }

        for entry in (pathVariable ?? "").split(separator: ":", omittingEmptySubsequences: true) {
            let dir = entry.trimmingCharacters(in: .whitespaces)
            if dir.isEmpty { continue }
            let candidate = (dir as NSString).appendingPathComponent("claude")
            if isExecutable(candidate) { return candidate }
        }
        return native
    }
}
